core/common/src/main/java/alluxio/collections/IndexedSet.java

```diff
@@ -128,21 +137,90 @@
   }
 
   /**
+   * An interface extending {@link FieldIndex}, represents a unique index. A Unique index is an
+   * index where each index value only maps to one object.
+   *
+   * @param <T> type of objects in this {@link IndexedSet}
+   */
+  public interface UniqueFieldIndex<T> extends FieldIndex<T> {
+  }
+
+  /**
+   * An interface extending {@link FieldIndex}, represents a non-unique index. A non-unique index is
+   * an index where a index value can map to one or more objects.
+   *
+   * @param <T> type of objects in this {@link IndexedSet}
+   */
+  public interface NonUniqueFieldIndex<T> extends FieldIndex<T> {
+  }
+
+  /**
    * Constructs a new {@link IndexedSet} instance with at least one field as the index.
    *
    * @param field at least one field is needed to index the set of objects
    * @param otherFields other fields to index the set
    */
   @SafeVarargs
   public IndexedSet(FieldIndex<T> field, FieldIndex<T>... otherFields) {
-    Map<FieldIndex<T>, ConcurrentHashMap<Object, ConcurrentHashSet<T>>> indexMap =
-        new HashMap<>(otherFields.length + 1);
-    indexMap.put(field, new ConcurrentHashMap<Object, ConcurrentHashSet<T>>(8, 0.95f, 8));
+    // count the numbers of two index types
```

Comment:
Start off by combining the indices into one Iterable with `Iterable<FieldIndex<T>> fields = Iterables.concat(Arrays.asList(field), Arrays.asList(otherFields)). This will make the rest of the code simpler.